### PR TITLE
ARROW-2585: [C++] Add Decimal::FromBigEndian, which was formerly a static method in parquet-cpp/src/parquet/arrow/reader.cc

### PR DESCRIPTION
**Pull request by joshuastorck**

### CODE REVIEW

File: cpp/src/arrow/util/decimal.h [hunk 1]
```diff
@@ -126,10 +126,18 @@ class ARROW_EXPORT Decimal128 {
   static Status FromString(const std::string& s, Decimal128* out,
                            int32_t* precision = NULLPTR, int32_t* scale = NULLPTR);
 
+  /// \brief Convert from a big endian byte representation. The length must be
+  ///        between 1 and 16
+  /// \return error status if the length is an invalid value
+  static Status FromBigEndian(const uint8_t* data, std::int32_t length, Decimal128* out);
```
Comment (pitrou) [inline cpp/src/arrow/util/decimal.h]: You shouldn't need `std::` in front of `int32_t`.

File: cpp/src/arrow/util/decimal.cc [hunk 2]
```diff
@@ -873,4 +874,114 @@ Status Decimal128::Rescale(int32_t original_scale, int32_t new_scale,
...
+
+  switch (length) {
+    case 0:
+      *out = 0;
+      break;
+    case 1:
+      *out = *bytes;
+      break;
+    case 2:
+      *out = BitUtil::FromBigEndian(*reinterpret_cast<const uint16_t*>(bytes));
```
Comment (pitrou) [inline cpp/src/arrow/util/decimal.cc]: This is potentially doing an unaligned access, which is forbidden on some architectures (SPARC for example, perhaps others).

File: cpp/src/arrow/util/decimal-test.cc [hunk 3]
```diff
@@ -380,4 +380,16 @@ TEST(Decimal128Test, TestNoDecimalPointExponential) {
   ASSERT_EQ(0, scale);
 }
 
+TEST(Decimal128Test, TestFromBigEndian) {
+  Decimal128 value(1);
```
Comment (pitrou) [inline cpp/src/arrow/util/decimal-test.cc]: Ideally you should also test non-trivial values (not one which only has only one bit set), and also negative values.

**Review by pitrou — CHANGES_REQUESTED** [review]
From an API point of view, there's a `Decimal128` constructor from little-endian bytes that only takes a `uint8_t*` (without a length), and a static method factory from big-endian bytes that takes a pointer and a length. Shouldn't we have something more consistent?

File: cpp/src/arrow/util/decimal.cc (hunk 2 above)
Comment (cpcloud) [reply cpp/src/arrow/util/decimal.cc]: Looks like for linux there's [`get_unaligned`/`put_unaligned`](https://www.kernel.org/doc/Documentation/unaligned-memory-access.txt). For MSVC, we could create a template function that creates a temporary struct and access a member.

Comment (pitrou) [reply cpp/src/arrow/util/decimal.cc]: I don't think we need such complication. Just read the memory byte per byte. Or, if you really insist on doing a single `int64_t` read, use `mempcy` into an `int64_t` local variable. Most compilers will optimize that into an unaligned access if the platform allows for it.

Comment (joshuastorck) [reply cpp/src/arrow/util/decimal.cc]: I looked at the generated assembly using the switch statement and it's producing a series of cmp/jne statements, so implementing it in a loop will have the same performance without the alignment issues. I will make the change and push it.

**Comment by pitrou** [comment]
@joshuastorck, thanks for the test additions. What do you think about the API inconsistency? It would deserve fixing before doing it has a compatibility cost, IMO.

**Comment by joshuastorck** [comment]
Since we are using Status codes (I wish we were using exceptions) for errors and this function can return an error, I don't want the logic in the constructor because the only way to handle that error is to do what's being done in the constructor that takes a string, which is to log and abort. 

If we were using exceptions, I would advocate for the same approach, primarily because this is a primitive numeric type and users don't want to think about exceptions when they are using primitive numeric types.

**Comment by pitrou** [comment]
That makes sense. I think we can go with this, then.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@joshuastorck, sorry, I forgot to mention: you should prefix this PR's title with the corresponding JIRA ticket number and category, like this: `ARROW-1234: [C++] Adding Decimal::FromBigEndian etc.`.

File: cpp/src/arrow/util/decimal.h [hunk 4]
```diff
@@ -126,10 +126,18 @@ class ARROW_EXPORT Decimal128 {
...
+  ///        between 1 and 16
+  /// \return error status if the length is an invalid value
+  static Status FromBigEndian(const uint8_t* data, int32_t length, Decimal128* out);
+
   /// \brief Convert Decimal128 from one scale to another
   Status Rescale(int32_t original_scale, int32_t new_scale, Decimal128* out) const;
 
  private:
+  // Helper function used by FromBigEndian that returns a Decimal128
+  static uint64_t FromBigEndian(const uint8_t* bytes, int32_t length);
```
[the remark this answers is not in the document]
Comment (joshuastorck) [reply cpp/src/arrow/util/decimal.h]: Done.

**Comment by joshuastorck** [comment]
@pitrou, I think I've resolved all of your concerns. Could you please merge?

**Comment by pitrou** [comment]
@joshuastorck I'll probably do it in a couple days. If there is any hurry, please tell me.

**Comment by joshuastorck** [comment]
It's not urgent, I just want to make sure that it makes it in so that I can
get the Travis tests working for this PR:
https://github.com/apache/parquet-cpp/pull/462.
